### ISSUE TITLE
Insufficient defense against concurrent operations on a backup repository

### ISSUE THREAD

**Issue by zjw** [rdiff-backup/rdiff-backup#122]
rdiff-backup blocks certain types of concurrent operations that could cause problems, such as the initiation of either a backup or a restore while a backup is already ongoing.  But there are a couple cases of concurrency that are not blocked, but should be.

It is currently possible (in v. 1.2.8) for multiple instances of `rdiff-backup --check-destination-dir` to be running simultaneously on a single backup repository.  I am not sure if this could lead to corruption of the repository, but at a minimum, it is a waste of processing time & resources.

A real-life scenario of this occurring was mentioned [here](https://lists.nongnu.org/archive/html/rdiff-backup-users/2019-09/msg00007.html).  In that particular instance, the user was performing a regression after a failed backup.  The regression was taking many hours (as is common for large repositories).  Meanwhile, the user's regularly scheduled daily backup automatically started, which initiated another regression of the repository, even while the earlier regression was still ongoing.

Proper behavior would be for a regression to terminate immediately upon detecting that another regression was already in progress.

Another case of concurrency that is not blocked is the initiation of a backup while a restore operation is already ongoing.  This type of concurrency should also be blocked.

**Comment by zjw** [comment]
Expanding on my original comments -- The first concurrency problem I mentioned can lead to a situation with a backup running concurrently with a regression:

1. Start a backup after a previously failed backup.  The failed backup is detected, and a regression is initiated.
2. Before the above regression is complete, start another backup, which initiates another regression.
3. Once the first regression is complete, the first backup begins, even while the second regression is ongoing.

Again, I don't know if this could lead to corruption.  If nothing else, the second regression will finish and will give up when attempting to remove a `current_mirror` file that was already removed when the first regression finished.

**Comment by ikus060** [comment]
In additional, the `current_mirror` file with a pid is not reliable. multiple rdiff-backup process might be running on different server with the same NFS mount point. Or running inside a docker container... PID is not the way to go.

Locking mechanism that is reliable is not simple if we want to make it work on all the file-system (local or network). File Lock is simple not supported properly on network file system.

The only real implementation I have experience with and that is working properly is H2DB FileLock:
http://www.h2database.com/html/advanced.html#file_locking_protocols

**Comment by ericzolf** [comment]
I've split out the more reliable approach over network file system topic into a separate issue (as enhancement) because it sounds much more difficult to realise than just making sure that check-destination-dir follows the current locking mechanism.

**Comment by ericzolf** [comment]
After a deeper look at the code, we have a few issues to address:

1. actually, we need to make the difference between "a backup has been interrupted" vs. "some action is already on-going"
2. in the first case, a regression is necessary (which itself is an action creating a lock on the repo)
3. in the 2nd case, whichever action has been requested can't proceed until the on-going one is finished (the real lock)
4. the current_mirror thingy actually combines both meanings which means that you can't notice that a regression is already on-going (the 2nd current_mirror has still an out-dated PID).
5. currently also restores would be possible while a regression is on-going, which is also dangerous, and restore isn't able to make the difference between an on-going backup and an interrupted one.
6. we could allow multiple restores in parallel, hence make the difference between write and read access, but I wouldn't want to add this level of complexity.
7. as a temporary workaround, it could be possible to create a 3rd current_mirror while regressing, I guess this would solve at least partially the issue but that wouldn't make the code easier to understand...

So, I hesitate if should go now for the workaround or wait until I've cleaned up the code, and solve the issue then one and for all.

**Comment by ikus060** [comment]
I'm getting hit by this issue real mad. I have a backup running in background every 24h in a cron. The backup take so long that I now have 3 rdiff-backup process running over and over again. One trying to backup, the second trying to regress and I have no idea what is the third is doing !

When we have two current_mirror, nothing seams really in place to check if the PID is running. rdiff-backup just assume the backup has been interrupted.

**Comment by ericzolf** [comment]
Agreed, that's part of the problem.
I've been looking a bit into it: https://www.oreilly.com/library/view/python-cookbook/0596001673/ch04s25.html but I'm still thinking about a cleaner connection interface before I implement something like this.
